plugins/secretstores/gdchauth/gdchauth.go

```diff
@@ -0,0 +1,276 @@
+package gdchauth
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/x509"
+	_ "embed"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/golang-jwt/jwt/v4"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/common/tls"
+	"github.com/influxdata/telegraf/plugins/secretstores"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+/* #nosec G101 */
+const (
+	tokenExchangeType       = "urn:ietf:params:oauth:token-type:token-exchange"
+	accessTokenTokenType    = "urn:ietf:params:oauth:token-type:access_token"
+	serviceAccountTokenType = "urn:k8s:params:oauth:token-type:serviceaccount"
+)
+
+func (*GdchAuth) SampleConfig() string {
+	return sampleConfig
+}
+
+// GdchAuth is the main authenticator struct
+type GdchAuth struct {
+	ServiceAccountFile string          `toml:"service_account_file"`
+	Audience           string          `toml:"audience"`
+	TokenExpiryBuffer  config.Duration `toml:"token_expiry_buffer"`
+	tls.ClientConfig   `toml:"tls"`
+
+	Log telegraf.Logger `toml:"-"`
+
+	token       string
+	tokenExpiry time.Time
+	tokenMutex  sync.RWMutex
+	saKey       *ServiceAccountKey
+	httpClient  *http.Client
+}
+
+type ServiceAccountKey struct {
+	PrivateKeyID        string `json:"private_key_id"`
+	PrivateKey          string `json:"private_key"`
+	ServiceIdentityName string `json:"name"`
+	TokenURI            string `json:"token_uri"`
+	Project             string `json:"project"`
+
+	parsedKey     crypto.Signer
+	signingMethod jwt.SigningMethod
+}
```

Comment:
How about moving this structure out into a `types.go` file to mostly keep the plugin code in there? Same for other custom received or sent types.